File: tool/gen_icon.py
#!/usr/bin/env python3
"""
LoadOut placeholder app icon generator.

Generates two PNGs into assets/icon/:
  - icon.png             1024x1024 master (solid background, edge-to-edge)
  - icon_foreground.png  1024x1024 with transparent background, motif at ~60%
                         scale (Android adaptive icon foreground; system fills
                         the background separately)

Design (placeholder):
  Headstamp-inspired motif. Dark gunmetal background, brass-colored outer ring
  and a serif "LO" wordmark in the center. Restrained, no bullets/skulls/etc.

Re-run any time with:
  python3 tool/gen_icon.py
"""

import os
import sys
from pathlib import Path

try:
    from PIL import Image, ImageDraw, ImageFont, ImageFilter
except ImportError:
    sys.stderr.write(
        "Pillow (PIL) is required.\n"
        "Install with: pip3 install --user --break-system-packages pillow\n"
    )
    raise

# ---------- Configuration -----------------------------------------------------

CANVAS = 1024  # px
PROJECT_ROOT = Path(__file__).resolve().parent.parent
OUT_DIR = PROJECT_ROOT / "assets" / "icon"

# Committed colors:
#   Background = charcoal gunmetal
#   Foreground = warm brass
BG_COLOR = (31, 41, 55, 255)        # #1F2937
FG_COLOR = (197, 165, 114, 255)     # #C5A572
FG_DIM = (165, 138, 95, 255)        # darker brass for inner stamp ring

# Font candidates, in priority order. PIL falls back if none are found.
FONT_CANDIDATES = [
    "/System/Library/Fonts/Supplemental/Times New Roman Bold.ttf",
    "/System/Library/Fonts/Supplemental/Georgia Bold.ttf",
    "/System/Library/Fonts/Supplemental/Times New Roman.ttf",
    "/System/Library/Fonts/Supplemental/Georgia.ttf",
    "/System/Library/Fonts/Times.ttc",
    "/System/Library/Fonts/Georgia.ttf",
    "/Library/Fonts/Times New Roman Bold.ttf",
    "/Library/Fonts/Georgia Bold.ttf",
]

WORDMARK = "LO"


# ---------- Helpers -----------------------------------------------------------

def find_font(size: int) -> tuple[ImageFont.FreeTypeFont | ImageFont.ImageFont, str]:
    """Return (font, path) for the first font candidate that loads."""
    for path in FONT_CANDIDATES:
        if os.path.isfile(path):
            try:
                return ImageFont.truetype(path, size=size), path
            except Exception:
                continue
    # Last-ditch fallback. Looks bad but lets the script complete.
    return ImageFont.load_default(), "<PIL default>"


def draw_motif(img: Image.Image, scale: float = 1.0) -> None:
    """
    Draw the headstamp motif onto img. The motif is sized relative to CANVAS;
    `scale` shrinks it (1.0 = full canvas, 0.6 = 60% for adaptive foreground).

    Motif: outer brass ring + thin inner ring (stamping shoulder) + bold serif
    "LO" wordmark centered.
    """
    cx = cy = CANVAS / 2
    motif_radius = (CANVAS / 2) * 0.86 * scale  # leave a little air at the edge

    draw = ImageDraw.Draw(img)

    # Outer ring stroke width scales with the motif.
    outer_stroke = max(2, int(motif_radius * 0.045))
    inner_stroke = max(1, int(motif_radius * 0.018))

    # Outer ring
    bbox_outer = (
        cx - motif_radius,
        cy - motif_radius,
        cx + motif_radius,
        cy + motif_radius,
    )
    draw.ellipse(bbox_outer, outline=FG_COLOR, width=outer_stroke)

    # Inner concentric ring (the "stamp shoulder")
    inner_radius = motif_radius * 0.78
    bbox_inner = (
        cx - inner_radius,
        cy - inner_radius,
        cx + inner_radius,
        cy + inner_radius,
    )
    draw.ellipse(bbox_inner, outline=FG_DIM, width=inner_stroke)

    # Wordmark
    # We size the font so the rendered text fits comfortably inside the inner
    # ring. Start big and shrink until it fits.
    target_text_height = inner_radius * 1.05  # bold and prominent
    font_size = int(target_text_height)
    while font_size > 10:
        font, font_path = find_font(font_size)
        bbox = draw.textbbox((0, 0), WORDMARK, font=font)
        w = bbox[2] - bbox[0]
        h = bbox[3] - bbox[1]
        # Limit width to ~1.35 * inner radius so it never crowds the ring.
        if w <= inner_radius * 1.45 and h <= inner_radius * 1.10:
            break
        font_size -= 8
    else:
        font, font_path = find_font(60)
        bbox = draw.textbbox((0, 0), WORDMARK, font=font)
        w = bbox[2] - bbox[0]
        h = bbox[3] - bbox[1]

    # Center the text. Account for the bbox offset (PIL textbbox can have a
    # non-zero top/left for many fonts).
    text_x = cx - (w / 2) - bbox[0]
    text_y = cy - (h / 2) - bbox[1]
    draw.text((text_x, text_y), WORDMARK, font=font, fill=FG_COLOR)

    # Stash font path on the image for the caller (best-effort report).
    img.info["__font_path"] = font_path


# ---------- Outputs -----------------------------------------------------------

def render_master() -> Path:
    """1024x1024 PNG, opaque dark background, motif at full scale."""
    img = Image.new("RGBA", (CANVAS, CANVAS), BG_COLOR)
    draw_motif(img, scale=1.0)
    out = OUT_DIR / "icon.png"
    # Flatten to RGB so iOS doesn't see any alpha channel.
    img.convert("RGB").save(out, format="PNG", optimize=True)
    return out


def render_adaptive_foreground() -> Path:
    """1024x1024 PNG, transparent background, motif at 60% scale (safe zone)."""
    img = Image.new("RGBA", (CANVAS, CANVAS), (0, 0, 0, 0))
    draw_motif(img, scale=0.60)
    out = OUT_DIR / "icon_foreground.png"
    img.save(out, format="PNG", optimize=True)
    return out


def main() -> int:
    OUT_DIR.mkdir(parents=True, exist_ok=True)

    master = render_master()
    fg = render_adaptive_foreground()

    # Tiny report to stdout.
    sample_font, font_path = find_font(200)
    print(f"Wrote {master} ({master.stat().st_size} bytes)")
    print(f"Wrote {fg}     ({fg.stat().st_size} bytes)")
    print(f"Background color: #{BG_COLOR[0]:02X}{BG_COLOR[1]:02X}{BG_COLOR[2]:02X}")
    print(f"Foreground color: #{FG_COLOR[0]:02X}{FG_COLOR[1]:02X}{FG_COLOR[2]:02X}")
    print(f"Font: {font_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: bin/gen_icon.dart
// LoadOut placeholder app icon generator (Dart edition).
//
// Generates two PNGs into assets/icon/:
//   - icon.png             1024x1024 master (solid background, edge-to-edge)
//   - icon_foreground.png  1024x1024 with transparent background, motif at
//                          ~60% scale (Android adaptive icon foreground;
//                          system fills the background separately)
//
// Design (placeholder):
//   Headstamp-inspired motif. Dark gunmetal background, brass-colored outer
//   ring + thinner inner stamp ring + bold geometric "LO" wordmark.
//   Restrained, no bullets/skulls/etc.
//
// The "LO" wordmark is drawn as pure primitives (rectangles + ring) instead
// of using a font. The image package's bitmap fonts blur badly when scaled
// to icon sizes, and it has no TrueType support. For a 2-letter placeholder
// glyph, hand-drawn primitives are crisper anyway.
//
// Re-run with:
//   dart pub run loadout:gen_icon
//
// Companion script tool/gen_icon.py exists for reference (the original
// design intent was Python+PIL with a real serif TTF), but this Dart port
// is what actually runs in the dev environment.

// CLI tool — `print` is the right output channel here. The `image` package is
// a dev_dependency used only by this build script, which the lint flags but
// is correct to import here.
// ignore_for_file: avoid_print, depend_on_referenced_packages

import 'dart:io';
import 'dart:math' as math;

import 'package:image/image.dart' as img;

const int canvas = 1024;

// Committed colors:
//   Background = charcoal gunmetal #1F2937
//   Foreground = warm brass        #C5A572
final bgColor = img.ColorRgb8(0x1F, 0x29, 0x37);
final fgColor = img.ColorRgb8(0xC5, 0xA5, 0x72);
final fgDim = img.ColorRgb8(0xA5, 0x8A, 0x5F);
final transparentRgba = img.ColorRgba8(0, 0, 0, 0);

/// Fill a circle with antialiasing using the package's primitive.
void disc(img.Image image, num cx, num cy, num r, img.Color color) {
  img.fillCircle(
    image,
    x: cx.round(),
    y: cy.round(),
    radius: r.round(),
    color: color,
    antialias: true,
  );
}

/// Draw an annulus (ring) directly by setting pixels. Each pixel's alpha is
/// derived from its signed distance to the centerline of the ring, giving a
/// soft anti-aliased edge that works on both opaque and transparent
/// canvases. We can't use disc-then-punch on an RGBA canvas because
/// fillCircle composites rather than replaces.
void ring(
  img.Image image, {
  required double cx,
  required double cy,
  required double radius,
  required double strokeWidth,
  required img.Color color,
}) {
  final outer = radius + strokeWidth / 2;
  final inner = radius - strokeWidth / 2;
  final cr = color.r.toInt();
  final cg = color.g.toInt();
  final cb = color.b.toInt();

  // Bounding box for the ring.
  final x0 = (cx - outer - 1).floor().clamp(0, image.width - 1);
  final x1 = (cx + outer + 1).ceil().clamp(0, image.width - 1);
  final y0 = (cy - outer - 1).floor().clamp(0, image.height - 1);
  final y1 = (cy + outer + 1).ceil().clamp(0, image.height - 1);

  for (var y = y0; y <= y1; y++) {
    final dy = y + 0.5 - cy;
    for (var x = x0; x <= x1; x++) {
      final dx = x + 0.5 - cx;
      final d = math.sqrt(dx * dx + dy * dy);
      // Alpha contribution = how much of the pixel is inside the ring.
      // For a ring, full alpha when inner < d < outer, with a 1px taper at
      // each edge.
      double a;
      if (d < inner - 0.5 || d > outer + 0.5) {
        continue;
      } else if (d > inner + 0.5 && d < outer - 0.5) {
        a = 1.0;
      } else if (d <= inner + 0.5) {
        a = (d - (inner - 0.5)).clamp(0.0, 1.0);
      } else {
        a = ((outer + 0.5) - d).clamp(0.0, 1.0);
      }
      _blendPixel(image, x, y, cr, cg, cb, a);
    }
  }
}

/// Composite one source pixel (RGB + alpha 0..1) onto [image] at (x, y),
/// preserving the channel count of the destination.
void _blendPixel(
  img.Image image,
  int x,
  int y,
  int sr,
  int sg,
  int sb,
  double srcA,
) {
  if (srcA <= 0) return;
  final p = image.getPixel(x, y);
  final dr = p.r.toInt();
  final dg = p.g.toInt();
  final db = p.b.toInt();
  final hasAlpha = image.numChannels == 4;
  final dAlpha = hasAlpha ? p.a / 255.0 : 1.0;

  // Standard "over" compositing.
  final outA = srcA + dAlpha * (1.0 - srcA);
  if (outA <= 0) return;
  final outR = (sr * srcA + dr * dAlpha * (1.0 - srcA)) / outA;
  final outG = (sg * srcA + dg * dAlpha * (1.0 - srcA)) / outA;
  final outB = (sb * srcA + db * dAlpha * (1.0 - srcA)) / outA;

  if (hasAlpha) {
    image.setPixelRgba(
      x,
      y,
      outR.round(),
      outG.round(),
      outB.round(),
      (outA * 255).round(),
    );
  } else {
    image.setPixelRgb(x, y, outR.round(), outG.round(), outB.round());
  }
}

/// Draw a filled rectangle (axis-aligned).
void rect(
  img.Image image, {
  required num x1,
  required num y1,
  required num x2,
  required num y2,
  required img.Color color,
}) {
  img.fillRect(
    image,
    x1: x1.round(),
    y1: y1.round(),
    x2: x2.round(),
    y2: y2.round(),
    color: color,
  );
}

/// Draw a chunky serif-ish "L" inside the rectangle (left, top, right,
/// bottom), in [color]. Hand-drawn geometry: a vertical stem with a serif
/// foot at the top, and a horizontal arm with a serif tip at the bottom-
/// right.
void drawL(
  img.Image image, {
  required double left,
  required double top,
  required double right,
  required double bottom,
  required img.Color color,
}) {
  final w = right - left;
  final h = bottom - top;

  // Vertical stem
  final stemW = w * 0.34;
  rect(
    image,
    x1: left,
    y1: top,
    x2: left + stemW,
    y2: bottom,
    color: color,
  );

  // Top serif (small overhang on both sides at top of stem)
  final topSerifH = h * 0.10;
  final topSerifOver = w * 0.13;
  rect(
    image,
    x1: left - topSerifOver,
    y1: top,
    x2: left + stemW + topSerifOver,
    y2: top + topSerifH,
    color: color,
  );

  // Bottom horizontal arm
  final armH = h * 0.22;
  rect(
    image,
    x1: left,
    y1: bottom - armH,
    x2: right,
    y2: bottom,
    color: color,
  );

  // Right tip serif (small upward foot at the end of the arm)
  final tipH = h * 0.12;
  final tipW = w * 0.06;
  rect(
    image,
    x1: right - tipW,
    y1: bottom - armH - tipH,
    x2: right,
    y2: bottom - armH,
    color: color,
  );
}

/// Draw an "O" centered on (cx, cy) as a thick ring of the given visual
/// height. Uses the direct-pixel ring helper (not disc-punch-disc) so it
/// works correctly on transparent canvases.
void drawO(
  img.Image image, {
  required double cx,
  required double cy,
  required double height,
  required img.Color color,
}) {
  final outerR = height / 2.0;
  final strokeW = outerR * 0.30;
  final centerR = outerR - strokeW / 2;

  ring(
    image,
    cx: cx,
    cy: cy,
    radius: centerR,
    strokeWidth: strokeW,
    color: color,
  );
}

/// Draw the LoadOut headstamp motif onto [image]. The motif is sized
/// relative to CANVAS; [scale] shrinks it (1.0 = full canvas, 0.6 = 60%
/// for adaptive foreground safe zone). Works on either an opaque canvas
/// (3-channel) or transparent canvas (4-channel).
void drawMotif(img.Image image, {required double scale}) {
  final cx = canvas / 2.0;
  final cy = canvas / 2.0;
  final motifRadius = (canvas / 2.0) * 0.86 * scale;

  final outerStroke = math.max(2.0, motifRadius * 0.045);
  final innerStroke = math.max(1.0, motifRadius * 0.020);

  // Outer ring (bold)
  ring(
    image,
    cx: cx,
    cy: cy,
    radius: motifRadius,
    strokeWidth: outerStroke,
    color: fgColor,
  );

  // Inner stamp ring (thin, slightly dimmer)
  final innerRingR = motifRadius * 0.78;
  ring(
    image,
    cx: cx,
    cy: cy,
    radius: innerRingR,
    strokeWidth: innerStroke,
    color: fgDim,
  );

  // "LO" wordmark sized to read clearly inside the inner ring.
  //
  // Layout: L on the left, O on the right, sharing a baseline. The L is
  // narrower than the O (typical serif proportion) so we size each glyph
  // separately. Total wordmark width = lWidth + spacing + oWidth, centered.
  final glyphHeight = innerRingR * 0.95;
  final lWidth = glyphHeight * 0.66;     // serif L is roughly 0.65-0.7 H
  final oWidth = glyphHeight;            // O is square (approximately)
  final spacing = glyphHeight * 0.09;
  final wordWidth = lWidth + spacing + oWidth;

  final wordLeft = cx - wordWidth / 2;
  final wordTop = cy - glyphHeight / 2;

  // L
  drawL(
    image,
    left: wordLeft,
    top: wordTop,
    right: wordLeft + lWidth,
    bottom: wordTop + glyphHeight,
    color: fgColor,
  );

  // O
  final oCx = wordLeft + lWidth + spacing + oWidth / 2;
  drawO(
    image,
    cx: oCx,
    cy: cy,
    height: glyphHeight,
    color: fgColor,
  );
}

Future<void> writePng(img.Image image, String path) async {
  final bytes = img.encodePng(image, level: 6);
  final file = File(path);
  await file.parent.create(recursive: true);
  await file.writeAsBytes(bytes);
  print('Wrote $path (${bytes.length} bytes)');
}

String hex(img.Color c) {
  String h(num v) => v.toInt().toRadixString(16).padLeft(2, '0').toUpperCase();
  return '#${h(c.r)}${h(c.g)}${h(c.b)}';
}

Future<void> main() async {
  // Master: solid dark background, full-scale motif. Opaque (numChannels: 3)
  // so iOS doesn't see any alpha channel.
  final master = img.Image(width: canvas, height: canvas, numChannels: 3);
  img.fill(master, color: bgColor);
  drawMotif(master, scale: 1.0);
  await writePng(master, 'assets/icon/icon.png');

  // Adaptive foreground: transparent canvas, motif at 60% scale. RGBA so we
  // get a real alpha channel.
  final fg = img.Image(width: canvas, height: canvas, numChannels: 4);
  img.fill(fg, color: transparentRgba);
  drawMotif(fg, scale: 0.60);
  await writePng(fg, 'assets/icon/icon_foreground.png');

  print('Background color: ${hex(bgColor)}');
  print('Foreground color: ${hex(fgColor)}');
  print('Glyphs: hand-drawn geometric "LO" wordmark (no font dependency).');
}
